sdk/go/common/tokens/stack_name.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tokens
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/env"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+// StackName is a valid stack name. It should always be initialised via ParseStackName, the use of it's zero
+// value will panic.
+type StackName struct {
+	str string
+}
+
+// IsEmpty returns true if the stack name is empty.
+func (sn StackName) IsEmpty() bool {
+	return sn.str == ""
+}
+
+// String implements fmt.Stringer. This method panics if StackName was zero initialized.
+func (sn StackName) String() string {
+	if env.DisableValidation.Value() {
+		contract.Assertf(sn.str != "", "stack name must not be empty")
+	}
+	return sn.str
+}
+
+// Q is a convenience method that returns the stack name as a QName. This method panics if StackName was zero
+// initialized.
+func (sn StackName) Q() QName {
+	return QName(sn.String())
+}
+
+var (
+	stackNameStartRegex = regexp.MustCompile("^[A-Za-z0-9]")
```

Comment:
This is a small behavior change to try and keep stacknames as something "identifier"-like. I've looked at a load of service data and no ones started their stacks in there with ".", "-", or "_" but there are a lot starting with digits.

I'm not totally sure we want to do this, might be better to be risk averse and allow those three characters as starting chars. Even if service data had no stacks using it we can't know for filestate.